packages/nimble/inst/tests/test_utils.R

```diff
@@ -26,6 +26,49 @@ temporarilyAssignInGlobalEnv <- function(value) {
     do.call('on.exit', list(rmCommand), envir = parent.frame())
 }
 
+withTempProject <- function(code) {
+    code <- substitute(code)
+    project <- nimble:::nimbleProjectClass()
+    nimbleOptions(nimbleProject = project)
+    on.exit({
+        ## messages are suppressed by test_that, so "assign('.check', 1, globalenv())" can be used as a way to verify this code was called 
+        .project <- nimbleOptions()$nimbleProject
+        nimbleOptions(nimbleProject = NULL) ## clear this before clearCompiled step, in case clearCompiled() itself fails
+        .project$clearCompiled()
+    })
+    try(eval(code, envir = parent.frame()))
```

Comment:
I think this would simpler using a `tryCatch(-, finally)` construct:
```
withTempProject <- function(code) {
    code <- substitute(code)
    project <- nimble:::nimbleProjectClass()
    nimbleOptions(nimbleProject = project)
    tryCatch(eval(code, envir = parent.frame()), finally = {
        ## messages are suppressed by test_that, so "assign('.check', 1, globalenv())" can be used as a way to verify this code was called 
        .project <- nimbleOptions()$nimbleProject
        nimbleOptions(nimbleProject = NULL) ## clear this before clearCompiled step, in case clearCompiled() itself fails
        .project$clearCompiled()
    })
}
```
This has different behavior than your first implementation, since it runs cleanup code without catching the exception. To observe the difference, note that in composite tests, the return value will be hidden:
```
test_that('function compiles and DSL version works', {
    nimFun <- nimbleFunction(run = function() { return(0); returnType(double(0)) })
    expect_compiles(nimFun)  # This can never fail in your version 1, but can fail in the tryCatch version.
    expect_equal(nimFun(), 0)
})
```
Thanks for taking the time to work this out! I'm definitely learning more about R's exception mechanism while trying to get this right.